### PR TITLE
[Merged by Bors] - feat(Topology/KrullDimension.lean): Krull dimension of a topological space

### PR DESCRIPTION
**Pull request by FMLJohn**
Define the Krull dimension of a topological space: the Krull dimension of a topological space is the supremum of lengths of chains of closed irreducible sets.

----------

@jjaassoonn

### CODE REVIEW

File: Mathlib/Topology/KrullDimension.lean [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+the length of longest series of closed irreducible subsets ordered by inclusion.
+-/
+
+/--
+The Krull dimension of a topological space is the supremum of lengths of chains of
+closed irreducible sets.
+-/
+noncomputable def topologicalKrullDim (T : Type _) [TopologicalSpace T] :
+    WithBot (WithTop ℕ) :=
+  krullDim { s : Set T | IsClosed s ∧ IsIrreducible s }
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/Topology/KrullDimension.lean:24]: I think yes. `irreducibleComponents_eq_maximals_closed` also mentions this.

File: Mathlib/Topology/KrullDimension.lean [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+-/
+import Mathlib.Order.KrullDimension
+import Mathlib.Topology.Irreducible
+
+/-!
+# The Krull dimension of a topological space
+
+The Krull dimension of a topological space is the order theoretic Krull dimension applied to the
+collection of all its subsets that are closed and irreducible. Unfolding this definition, it is
+the length of longest series of closed irreducible subsets ordered by inclusion.
```
Comment (kbuzzard) [inline Mathlib/Topology/KrullDimension.lean:14]: Why not add some TODOs? For example Krull dimension of Spec(R) is Krull dimension of R, for R a commutative ring (do we have Krull dimension of a commutative ring?)

File: Mathlib/Topology/KrullDimension.lean (hunk 1 above)
[the remark this answers is not in the document]
Comment (kbuzzard) [reply Mathlib/Topology/KrullDimension.lean]: So does `vanishingIdeal_isClosed_isIrreducible` (this time in the other order). This proof is only used once in mathlib, in the proof of `vanishingIdeal_irreducibleComponents` (which is now worse than it was). To avoid this regression probably `vanishingIdeal_isClosed_isIrreducible` should be renamed `vanishingIdeal_isIrreducibleClosed` and fixed up to use this new concept.

File: Mathlib/Topology/KrullDimension.lean [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+-/
+
+/--
+The Krull dimension of a topological space is the supremum of lengths of chains of
+closed irreducible sets.
+-/
+noncomputable def topologicalKrullDim (T : Type _) [TopologicalSpace T] :
+    WithBot (WithTop ℕ) :=
+  krullDim { s : Set T | IsIrreducibleClosed s }
```
Comment (jjaassoonn) [inline Mathlib/Topology/KrullDimension.lean:23]: Should we define `IrreducibleCloseds` as a subtype and set up an order on `IrreducibleCloseds` and define `topologicalKrullDim` as `krullDim( irreducibleCloseds T)`?

File: Mathlib/Topology/KrullDimension.lean (hunk 1 above)
[the remark this answers is not in the document]
Comment (FMLJohn) [reply Mathlib/Topology/KrullDimension.lean]: > So does `vanishingIdeal_isClosed_isIrreducible` (this time in the other order). This proof is only used once in mathlib, in the proof of `vanishingIdeal_irreducibleComponents` (which is now worse than it was). To avoid this regression probably `vanishingIdeal_isClosed_isIrreducible` should be renamed `vanishingIdeal_isIrreducibleClosed` and fixed up to use this new concept.

I have made changes according to your advice. Consequently, the names of some results have been changed.

File: Mathlib/Topology/KrullDimension.lean (hunk 3 above)
Comment (kbuzzard) [reply Mathlib/Topology/KrullDimension.lean:23]: Whether or not we do this, we have `IsIrreducible` and `IsClosed` so I don't see any harm in `IsIrreducibleClosed` for now. @erdOne do you have an opinion on this? If we're happy with the definition as it is, this is probably ready to go now. But if we want to change, probably now is a good time to do it.

File: Mathlib/Topology/KrullDimension.lean (hunk 1 above)
[the remark this answers is not in the document]
Comment (mariainesdff) [reply Mathlib/Topology/KrullDimension.lean]: I have added a structure `IrreducibleCloseds` and refactored some declarations to use it.

File: Mathlib/Topology/KrullDimension.lean (hunk 2 above)
Comment (mariainesdff) [reply Mathlib/Topology/KrullDimension.lean:14]: I added the TODO comment. That result is proven in #6277, which is too big, but @FMLJohn  is working on splitting it into smaller PRs.

File: Mathlib/Topology/KrullDimension.lean (hunk 3 above)
Comment (mariainesdff) [reply Mathlib/Topology/KrullDimension.lean]: I added that definition.

File: Mathlib/Topology/Sets/Closeds.lean [hunk 4]
```diff
@@ -366,4 +366,62 @@ protected lemma mem_prod {s : Clopens α} {t : Clopens β} {x : α × β} :
...
+
+@[ext]
+protected theorem ext {s t : IrreducibleCloseds α} (h : (s : Set α) = t) : s = t :=
+  SetLike.ext' h
+
+@[simp]
+theorem coe_mk (s : Set α) (h) : (mk s h : Set α) = s :=
+  rfl
+
+instance : LE (IrreducibleCloseds α) where
```
Comment (erdOne) [inline Mathlib/Topology/Sets/Closeds.lean:401]: Does this not follow from `SetLike` via typeclass inference?

File: Mathlib/Topology/Sober.lean [hunk 5]
```diff
@@ -179,6 +180,24 @@ noncomputable def irreducibleSetEquivPoints [QuasiSober α] [T0Space α] :
     simp [hs.2.closure_eq, ht.2.closure_eq]
 #align irreducible_set_equiv_points irreducibleSetEquivPoints
 
+/-- The closed irreducible subsets of a sober space bijects with the points of the space. -/
+noncomputable def irreducibleSetEquivPoints' [QuasiSober α] [T0Space α] :
```
Comment (erdOne) [inline Mathlib/Topology/Sober.lean:184]: I think you should just replace `irreducibleSetEquivPoints`.

File: Mathlib/Topology/KrullDimension.lean [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Fangming Li
+-/
+import Mathlib.Order.KrullDimension
+import Mathlib.Topology.Sets.Closeds
+
+/-!
+# The Krull dimension of a topological space
+
+The Krull dimension of a topological space is the order theoretic Krull dimension applied to the
```
Comment (kbuzzard) [inline Mathlib/Topology/KrullDimension.lean:12]:
```suggestion
The Krull dimension of a topological space is the order-theoretic Krull dimension applied to the
```

File: Mathlib/Topology/Sets/Closeds.lean [hunk 7]
```diff
@@ -366,4 +366,59 @@ protected lemma mem_prod {s : Clopens α} {t : Clopens β} {x : α × β} :
...
 end Clopens
 
+/-! ### Irreducible closed sets -/
+
+/-- The type of irreducible closed subsets of a topological space. -/
+structure IrreducibleCloseds (α : Type*) [TopologicalSpace α] where
+  /-- the carrier set, i.e. the points in this set -/
+  carrier : Set α
+  irreducibleClosed' : IsIrreducibleClosed carrier
+namespace IrreducibleCloseds
```
Comment (kbuzzard) [inline Mathlib/Topology/Sets/Closeds.lean:376]:
```suggestion
  irreducibleClosed' : IsIrreducibleClosed carrier

namespace IrreducibleCloseds
```

File: Mathlib/Topology/Sets/Closeds.lean [hunk 8]
```diff
@@ -366,4 +366,59 @@ protected lemma mem_prod {s : Clopens α} {t : Clopens β} {x : α × β} :
...
+
+@[ext]
+protected theorem ext {s t : IrreducibleCloseds α} (h : (s : Set α) = t) : s = t :=
+  SetLike.ext' h
+
+@[simp]
+theorem coe_mk (s : Set α) (h) : (mk s h : Set α) = s :=
+  rfl
+
+/-- The term of `TopologicalSpace.Closeds α` corresponding to a singleton. -/
```
Comment (kbuzzard) [inline Mathlib/Topology/Sets/Closeds.lean:401]:
```suggestion
/-- The term of `TopologicalSpace.IrreducibleCloseds α` corresponding to a singleton. -/
```

**Review by kbuzzard — APPROVED** [review]
This looks fine to me. I left some minor comments.

File: Mathlib/Topology/KrullDimension.lean [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+collection of all its subsets that are closed and irreducible. Unfolding this definition, it is
+the length of longest series of closed irreducible subsets ordered by inclusion.
+
+TODO: The Krull dimension of `Spec(R)` equal to the Krull dimension of `R`, for `R` a commutative
```
Comment (erdOne) [inline Mathlib/Topology/KrullDimension.lean:16]:
```suggestion
TODO: The Krull dimension of `Spec(R)` equals the Krull dimension of `R`, for `R` a commutative
```

File: Mathlib/Topology/Sober.lean [hunk 10]
```diff
@@ -166,18 +167,21 @@ attribute [local instance] specializationOrder
...
+    simp only [IsIrreducible.genericPoint_closure_eq, TopologicalSpace.IrreducibleCloseds.coe_mk,
+      closure_eq_iff_isClosed.mpr s.2.2]
+    rfl
   right_inv x := isIrreducible_singleton.closure.genericPoint_spec.eq
       (by rw [closure_closure]; exact isGenericPoint_closure)
   map_rel_iff' := by
     rintro ⟨s, hs⟩ ⟨t, ht⟩
     refine specializes_iff_closure_subset.trans ?_
     simp [hs.2.closure_eq, ht.2.closure_eq]
-#align irreducible_set_equiv_points irreducibleSetEquivPoints
```
Comment (erdOne) [inline Mathlib/Topology/Sober.lean:180]: Why is this removed?

File: Mathlib/Topology/Sets/Closeds.lean [hunk 11]
```diff
@@ -366,4 +366,59 @@ protected lemma mem_prod {s : Clopens α} {t : Clopens β} {x : α × β} :
...
+@[simps]
+def singleton [T1Space α] (x : α) : IrreducibleCloseds α :=
+  ⟨{x}, ⟨isIrreducible_singleton, isClosed_singleton⟩⟩
+
+@[simp] lemma mem_singleton [T1Space α] {a b : α} : a ∈ singleton b ↔ a = b := Iff.rfl
+
+/-- Equivalence between `IrreducibleCloseds α` and `{x : Set α // IsIrreducibleClosed x }`. -/
+@[simps apply symm_apply]
+def equivSubtype : IrreducibleCloseds α ≃ {x : Set α // IsIrreducibleClosed x } where
```
Comment (erdOne) [inline Mathlib/Topology/Sets/Closeds.lean:410]:
```suggestion
def equivSubtype : IrreducibleCloseds α ≃ { x : Set α // IsIrreducibleClosed x } where
```

File: Mathlib/Topology/Sets/Closeds.lean [hunk 12]
```diff
@@ -366,4 +366,59 @@ protected lemma mem_prod {s : Clopens α} {t : Clopens β} {x : α × β} :
...
+  toFun a   := ⟨a.1, a.2⟩
+  invFun a  := ⟨a.1, a.2⟩
+  left_inv  := fun ⟨_, _⟩ => rfl
+  right_inv := fun ⟨_, _⟩ => rfl
+
+variable (α) in
+/-- The equivalence `IrreducibleCloseds α ≃ {x : Set α // IsIrreducibleClosed x }` is an order
+isomorphism.-/
+def orderIsoSubtype : IrreducibleCloseds α ≃o {x : Set α // IsIrreducibleClosed x } :=
```
Comment (erdOne) [inline Mathlib/Topology/Sets/Closeds.lean:419]:
```suggestion
def orderIsoSubtype : IrreducibleCloseds α ≃o { x : Set α // IsIrreducibleClosed x } :=
```

File: Mathlib/Topology/Sets/Closeds.lean [hunk 13]
```diff
@@ -366,4 +366,59 @@ protected lemma mem_prod {s : Clopens α} {t : Clopens β} {x : α × β} :
...
+/-- Equivalence between `IrreducibleCloseds α` and `{x : Set α // IsIrreducibleClosed x }`. -/
+@[simps apply symm_apply]
+def equivSubtype : IrreducibleCloseds α ≃ {x : Set α // IsIrreducibleClosed x } where
+  toFun a   := ⟨a.1, a.2⟩
+  invFun a  := ⟨a.1, a.2⟩
+  left_inv  := fun ⟨_, _⟩ => rfl
+  right_inv := fun ⟨_, _⟩ => rfl
+
+variable (α) in
+/-- The equivalence `IrreducibleCloseds α ≃ {x : Set α // IsIrreducibleClosed x }` is an order
```
Comment (erdOne) [inline Mathlib/Topology/Sets/Closeds.lean:417]:
```suggestion
/-- The equivalence `IrreducibleCloseds α ≃ { x : Set α // IsIrreducibleClosed x }` is an order
```

File: Mathlib/Topology/Sober.lean (hunk 10 above)
Comment (FMLJohn) [reply Mathlib/Topology/Sober.lean:180]: I have recovered the #align.

File: Mathlib/Topology/Irreducible.lean [hunk 14]
```diff
@@ -114,17 +116,40 @@ theorem isClosed_of_mem_irreducibleComponents (s) (H : s ∈ irreducibleComponen
...
-theorem irreducibleComponents_eq_maximals_closed (X : Type*) [TopologicalSpace X] :
-    irreducibleComponents X = maximals (· ≤ ·) { s : Set X | IsClosed s ∧ IsIrreducible s } := by
+section IsIrreducibleClosed
+
+variable {S : Type*} [TopologicalSpace S] (s : Set S)
+
+/--
+A set `s : Set X` satisfies `IsIrreducibleClosed s` if `s` is irreducible and closed.
+-/
+def IsIrreducibleClosed : Prop := IsIrreducible s ∧ IsClosed s
```
Comment (kbuzzard) [inline Mathlib/Topology/Irreducible.lean:126]: Now you have the `IrreducibleCloseds` structure, this `def` should probably be removed.

**Comment by kbuzzard** [comment]
Thanks a lot!

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge
